src/ui/VirtualFileRenderer/VirtualFileRenderer.tsx

```diff
@@ -0,0 +1,323 @@
+import { useWindowVirtualizer } from '@tanstack/react-virtual'
+// eslint-disable-next-line no-restricted-imports
+import type { Dictionary } from 'lodash'
+import Highlight, { defaultProps } from 'prism-react-renderer'
+import { useEffect, useLayoutEffect, useRef } from 'react'
+import { useHistory, useLocation } from 'react-router-dom'
+
+import { requestAnimationTimeout } from 'shared/utils/animationFrameUtils'
+import { cn } from 'shared/utils/cn'
+import { prismLanguageMapper } from 'shared/utils/prismLanguageMapper'
+
+import { ColorBar } from './ColorBar'
+
+// copied from prism-react-renderer since they don't export it
+type Token = {
+  types: string[]
+  content: string
+  empty?: boolean
+}
+
+interface CodeBodyProps {
+  tokens: Token[][]
+  getLineProps: Highlight['getLineProps']
+  getTokenProps: Highlight['getTokenProps']
+  coverage: Dictionary<'H' | 'M' | 'P'>
+  codeDisplayOverlayRef: React.RefObject<HTMLDivElement>
+}
+
+const CodeBody = ({
+  tokens,
+  getLineProps,
+  getTokenProps,
+  coverage,
+  codeDisplayOverlayRef,
+}: CodeBodyProps) => {
+  const history = useHistory()
+  const location = useLocation()
+
+  const initialRender = useRef(true)
+  const wrapperRef = useRef<HTMLDivElement>(null)
+
+  // virtualize code lines
+  const virtualizer = useWindowVirtualizer({
+    count: tokens.length,
+    estimateSize: () => 18,
+    overscan: 75,
+    scrollMargin: codeDisplayOverlayRef.current?.offsetTop ?? 0,
+  })
+
+  const div = codeDisplayOverlayRef.current
+  if (div) {
+    // set the parent div height to the total size of the virtualizer
+    div.style.height = `${virtualizer.getTotalSize()}px`
+    div.style.position = 'relative'
+  }
```

Comment:
Does this have to get called on each render? Could is just be done once when the tokens are rendered? Idk how this works, but I assume that as many lines are rendered, this method (depending on the implementation) might become slower

Reply:
So this `getTotalSize` seems to just be the number of rows * estimate size so that it can set the correct height to expand the window so that it feels like a full window scroll, I don't think there's any major perf issues here, this is fairly similar as to how they have it in their example in the documentation, however they have the style being directly applied to the component, however because of our code highlighting setup that's a bit trickier to do, I could possibly refactor things but that would be a fairly big refactor at this point I believe.

Reply:
So i was a bit off on how this method actually works, however diving deeper it seems that most of the most complex calculations of this function are memoized after their first run so the cost of running this function goes down and I believe fairly performant.

- `getTotalSize`: https://github.com/TanStack/virtual/blob/main/packages/virtual-core/src/index.ts#L1001-L1019
- `getMeasurements`: https://github.com/TanStack/virtual/blob/main/packages/virtual-core/src/index.ts#L585